scipy/spatial/_ckdtree.pyx

```diff
@@ -528,13 +529,14 @@ cdef class cKDTree:
         def __get__(cKDTree self):
             cdef cKDTreeNode n
             cdef ckdtree *cself = self.cself
-            if self._python_tree is not None:
-                return self._python_tree
-            else:
-                n = cKDTreeNode()
-                n._setup(self, node=cself.ctree, level=0)
-                self._python_tree = n
-                return self._python_tree
+            with self._tree_lock:
+                if self._python_tree is not None:
+                    return self._python_tree
+                else:
+                    n = cKDTreeNode()
+                    n._setup(self, node=cself.ctree, level=0)
+                    self._python_tree = n
+                    return self._python_tree
```

Comment:
Can we use double-checked locking here? e.g.
```suggestion
            if self._python_tree is not None:
                return self._python_tree
                
            with self._tree_lock:
                # may have been assigned by another thread
                if self._python_tree is not None:
                    return self._python_tree
                else:
                    n = cKDTreeNode()
                    n._setup(self, node=cself.ctree, level=0)
                    self._python_tree = n
                    return self._python_tree
```

Reply:
This is Cython  not Python code, right? I don't think the double-checked locking is thread-safe without making `self._python_tree` an atomic variable, and I don't think Cython syntax supports that.

Reply:
AFAIK Cython does have atomics `libcpp` but I guess it would be a pain to have an atomic object. Oh well, this isn't the most critical for performance anyway so I guess we can just take the hit,

Reply:
@peterbell10 mentioned to me just now that he measured the performance impact of locking/unlocking and it's ~100 ns, which should be fine.

@andfoy can you revert the change?